### ISSUE TITLE
Change Request: Introduce a system to suppress existing errors

### ISSUE THREAD

**Issue by jfmengels** [eslint/eslint#16755]
### ESLint version

v8.31.0

### What problem do you want to solve?

I would like there to be a system to suppress existing errors from being reported, so that I can enable a rule as `"error"` and be notified when new ones show up.

I believe that the way that this is done today with ESLint and many other linters is through different imperfect means.

### Using warnings

This use-case is apparently what the [`"warn"` severity level is for](https://twitter.com/slicknet/status/1579527957491564544).

A large problem with warnings is that as soon as there are more than a few warnings, you don't notice new ones showing up. A common practice I've seen quite often is to avoid warnings altogether, and to only use errors to avoid new problems creeping in. But that doesn't solve the problem of all the existing errors.

Also, users can too easily ignore the new errors, so in a way, the rule is enabled without being enforced when IMO the point of a linter is to enforce rules.

### Using disable comments

One can use disable comments to temporarily suppress errors, by adding a comment like `// eslint-disable rule-name  -- FIX THIS LATER`

Disable comments can be used to enable a rule as an error early, by adding them everywhere where an error is currently reported (and that is actually something that can be automated by some linters).

But disable comments have the tendency to be hard to distinguish from other disable comments created for reasons such as false positives or disagreements on the rule, especially when there is no enforced need to add a message on the comment. Meaning that once you decide to tackle the existing errors, they can be hard to detect (or to distinguish from ones that are disabled for good reasons).

They also "pollute" the codebase in a way that is quite visible, and makes users numb to the fact of using disable comments.

### Ignoring parts of the project

It is also possible to simply disable the rule in each file that is currently reporting errors, either through manually configuring the rule in the ESLint config, or by adding a disable comment at the top of the file that disables the rule for the entire file.

This has multiple downsides:
- While new errors are enforced in the other files, new errors can creep in the ignored files
- If/when the errors in the ignored files get removed, the user has to remember to re-enable the rule on this file. Otherwise new errors can creep in also.

### What do you think is the correct solution?

Here is some prior art on suppressing existing errors:
- [`elm-review`](https://elm-review.com)'s suppression system - described in this article https://jfmengels.net/stop-the-bleed/
- [Betterer](https://phenomnomnominal.github.io/betterer/)
- [`ember-template-lint`'s TODO system](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/todos.md)

I am the author of the linter and blog post mentioned in the first bullet point. My recommendation would be to create a similar system to this one. It has worked really well so far for my users and I.
The blog post also describes the problems I've mentioned in greater detail than in this issue.

I am not willing to submit a PR nor a RFC as I am not using ESLint anymore and this would be a large endeavor, but I'd be happy to give advice and recommendations on how such a system should be put in place.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Thanks for all the links to other ways of suppressing errors. I think this is worth investigating some more, especially as part of the rewrite that's coming.

**Comment by sam3k** [comment]
Should there be a RFC for this proposal?

**Comment by nzakas** [comment]
Yes, we'll need an RFC for this, but we are going to hold off and think about this as part of the rewrite.

**Comment by sambev** [comment]
Fwiw. Rubocop also does something like this by generating a "todo" file. I think from a high level it generates a yaml file that essentially maps files to rules that are ignored for the listed files. That way other files and new files will still error for said rule.

**Comment by elwayman02** [comment]
Some additional prior art: 

ember-template-lint's TODO system, as mentioned above by @jfmengels, implements a very nice workflow for temporarily suppressing existing errors, with an eye on encouraging teams to fix them over time by eventually making them errors again with a phased, configurable approach. That functionality is implemented on top of a set of shared utils extracted for use outside ember-template-lint: https://github.com/lint-todo/utils

This system has thus been ported to eslint and stylelint through the use of custom formatters augmenting the output of the linter through these utils:

https://github.com/lint-todo/eslint-formatter-todo
https://github.com/lint-todo/stylelint-formatter-todo

However, `eslint-formatter-todo` does not work past eslint v8.24.0, due to changes that took place in the linter. I'm not sure what exactly changed myself, but the formatter approach is certainly brittle in this regard. It would be great to see this proposal move forward such that there could be built-in support for this workflow within eslint itself, in a way that's consistent across multiple linters popular in the JavaScript ecosystem.

**Comment by nzakas** [comment]
@elwayman02 thanks for the additional information, that's very helpful.

We are waiting on someone to write up an RFC for this feature, as the core team doesn't have time to look into this in the near future. If you're interested, the best way to move this forward is by putting together an RFC.

**Comment by scalvert** [comment]
@nzakas as the original author (with my close co-collaborator @rwjblue) of the TODO system, I'd be happy to take on writing an RFC that would propose integrating this capability into eslint. Let me try to carve out some time to write this up in the near future.

**Comment by nzakas** [comment]
@scalvert awesome! We'd really appreciate it. ❤️

**Comment by softius** [comment]
My workaround for this is using a standalone script that takes the JSON formatted output from ESLint, and use it to either generate a baseline file or re-evaluate the error messages against the baseline. The baseline file is a JSON file that contains the number of errors for each rule in each file. 

From my point of view, the baseline approach is ideal because it doesn't pollute the codebase with a large number of `// eslint-disable-next-lint ...` commands , and it makes it easier to control the outstanding items. Plus, it is easier & cleaner to regenerate the baseline file if you are adding rules gradually, instead of having to parse and adjust the codebase.

I would be more than happy to work on RFC or PR to do include this as part of `eslint` given there is enough interest for this approach from the core team and community.

**Comment by octogonz** [comment]
[@rushstack/eslint-patch](https://www.npmjs.com/package/@rushstack/eslint-patch#eslint-bulk-suppressions-feature) solves this using the [eslint-bulk](https://www.npmjs.com/package/@rushstack/eslint-bulk) CLI. We planned to propose integrating it directly into official ESLint eventually. The feature design is mostly stabilized, however the `scopeId` spec will not be finalized until we complete a companion feature for bulk suppressions of `tsc`/**tsconfig.json** errors (coming this summer).

**Comment by nzakas** [comment]
The RFC has been accepted:
https://github.com/eslint/rfcs/pull/119

**Comment by bartocc** [comment]
@jfmengels I think you could also mention https://github.com/mansona/lint-to-the-future by @mansona in your paragraph `What do you think is the correct solution?`

**Comment by lishaduck** [comment]
I believe this has landed now? (#19159)

**Comment by nzakas** [comment]
You're correct, looks like we just missed adding the note in the commit message.
